pkg/agent/controller/networkpolicy/allocator.go

```diff
@@ -98,3 +166,22 @@ func (a *idAllocator) release(id uint32) error {
 	a.availableSlice = append(a.availableSlice, id)
 	return nil
 }
+
+// asyncRuleCache maintains asynchronous network policy rule cache with ID as the key.
+type asyncRuleCache struct {
+	rules       cache.Store
+	deleteQueue workqueue.RateLimitingInterface
+}
+
+// asyncRuleCacheKeyFunc knows how to get key of a *rule.
+func asyncRuleCacheKeyFunc(obj interface{}) (string, error) {
+	rule := obj.(*types.PolicyRule)
+	return strconv.Itoa(int(rule.FlowID)), nil
+}
+
+func NewAsyncRuleCache() *asyncRuleCache {
+	return &asyncRuleCache{
+		rules:       cache.NewStore(asyncRuleCacheKeyFunc),
+		deleteQueue: workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "async_delete_networkpolicyrule"),
```

Comment:
I think a DelayingQueue is good enough? @tnqn

Reply:
I considered the scenario, where all the rules in the agent are deleted at once. In that case, rate-limiting at the queue in the network policy controller is probably good enough and does not need any more rate-limiting. I can move to DelayingQueue. Would like to hear Quan's opinion too. Thanks.

Reply:
If we only expect a delay when adding a key, I agree DelayingQueue is more suitable, to save unnecessary rate limiting calculation overhead.